### PR TITLE
feat(cli): --filter flag

### LINKED ISSUE #2232: Support `bun run --workspace`
### What version of Bun is running?

0.5.7

### What platform is your computer?

Darwin 22.3.0 arm64 arm

### What steps can reproduce the bug?

In a monorepo with npm workspaces, scripts defined in a specific package's `package.json` can be run with 

```sh
npm run my-script --workspace packages/my-package
```

### What is the expected behavior?

Bun should allow me to run the package's script via the `--workspace` arg.

A possible workaround is:

```sh
(cd packages/my-package && bun run my-script)
```

### What do you see instead?

error: missing script "my-script"

### Additional information

_No response_

### LINKED ISSUE #5207: Support `bun run --workspaces`
### What is the problem this feature would solve?

Run a package.json script in all workspaces (if it is defined)

### What is the feature you are proposing to solve the problem?

---

### What alternatives have you considered?

_No response_

### PR DESCRIPTION
**Pull request by gvilums**
### What does this PR do?

Adds support for the `--filter <PATTERN>` argument for `bun run`. See the [pnpm docs](https://pnpm.io/filtering#--filter-package_name) for details.

Note that currently only `--filter <name_pattern>` and `--filter ./<dir_pattern>` are supported. Selecting dependencies or dependents of packages is not yet possible.

Resolves #2232
Resolves #5207 

### How did you verify your code works?
Tests in `test/cli/run/filter-workspace.test.ts`.

### CODE REVIEW

File: src/cli/run_command.zig
Comment (dylan-conway) [inline src/cli/run_command.zig]:
```suggestion
            switch (bun.sys.chdir(olddir)) {
                .err => |err| {
                    ...
                },
                .success => {},
            }
```

File: src/cli/run_command.zig
Comment (dylan-conway) [inline src/cli/run_command.zig]: `bun.sys.chdir` here too

File: src/cli/filter_arg.zig
Comment (dylan-conway) [inline src/cli/filter_arg.zig]:
```suggestion
        var buf2: bun.PathBuffer = undefined;
```
is the `+ 1` needed? null byte should be included

File: src/cli.zig
Comment (dylan-conway) [inline src/cli.zig]: also `bun.sys.chdir`

File: test/cli/run/filter-workspace.test.ts
Comment (dylan-conway) [inline test/cli/run/filter-workspace.test.ts]: these console logs can be removed?

**Review by dylan-conway — COMMENTED** [review]
Looks good so far. One thing that stands out as a difference from pnpm: `pnpm run --filter="*" debug` will run all `debug` scripts including one from the root package

File: src/cli/filter_arg.zig
Comment (gvilums) [reply src/cli/filter_arg.zig]: updated to use PathBuffer

File: test/cli/run/filter-workspace.test.ts
Comment (gvilums) [reply test/cli/run/filter-workspace.test.ts]: removed

**Comment by gvilums** [comment]
Just pushed a commit that includes the root package in script resolution

File: src/cli/filter_arg.zig
Comment (dylan-conway) [inline src/cli/filter_arg.zig]: this glob walker should ignore `node_modules` and other directories that pnpm ignores if they exist. I believe this is done through the `ignore_filter_fn`

File: src/cli/filter_arg.zig
Comment (dylan-conway) [inline src/cli/filter_arg.zig]: package jsons found here may or may not actually be workspace packages. If the script isn't found in the `package.json` we shouldn't print an error
Screenshot 2024-01-17 at 4 06 37 PM

**Comment by Jarred-Sumner** [comment]
Running `bun-debug run --filter="*" echo hi` crashes

image

File: src/cli/filter_arg.zig
Comment (Jarred-Sumner) [inline src/cli/filter_arg.zig:98]: This duplicates memory twice.

File: src/cli/filter_arg.zig
Comment (Jarred-Sumner) [inline src/cli/filter_arg.zig]: In large directory trees, this approach takes seconds to show any output. We cannot ship it without being an iterator.

**Review by Jarred-Sumner — CHANGES_REQUESTED** [review]
4 issues before this can be shipped.

1) This approach is not fast enough. It needs to be an iterator so that the developer gets immediate feedback instead of waiting for potentially their entire directory tree to parse every package.json file.

2) There is a resource leak. After enough scripts are run, spawning any process fails with `SystemResources` which usually means a file descriptor leak.

3) it scans node_modules. It should not scan node_modules, .git, or any `.` directories.

4) We need tests for what happens when you run multiple JavaScript/TypeScript files in different projects same process (`bun run --filter="*" index.js`). Currently, Bun assumes only one JavaScript file every executes for a given thread. That assumption is not true when using these code paths.

5) It would be best to use the module resolution cache so that when running a JavaScript file later, the subsequent calls to getdents64() get cached. This will involve a small change to the GlobWalker.

File: src/cli/filter_arg.zig
Comment (gvilums) [reply src/cli/filter_arg.zig]: The `--if-present` flag accomplishes this, but maybe this should be the default, just like for pnpm.

File: src/cli/filter_arg.zig
Comment (gvilums) [reply src/cli/filter_arg.zig:98]: Does it? From my understanding we dupe the name (part of the hash map value) in the first line, and duple the path (hash map key) in the second line.

Comment (Jarred-Sumner) [reply src/cli/filter_arg.zig:98]: Right here:

https://github.com/oven-sh/bun/blob/52894121ea2436481b97b5318d1ce64176b18c0e/src/install/lockfile.zig#L3562-L3573

File: src/cli/filter_arg.zig
Comment (gvilums) [reply src/cli/filter_arg.zig]: understood, will get on that 👍

File: src/cli/filter_arg.zig
Comment (gvilums) [reply src/cli/filter_arg.zig]: fixed in latest commit

File: src/cli/run_command.zig
Comment (Jarred-Sumner) [inline src/cli/run_command.zig]: only this one should be necessary

**Comment by gvilums** [comment]
> 4 issues before this can be shipped.
> 
> 1. This approach is not fast enough. It needs to be an iterator so that the developer gets immediate feedback instead of waiting for potentially their entire directory tree to parse every package.json file.

This is now implemented in the `PackageFilterIterator ` in `filter_args.zig`
> 2. There is a resource leak. After enough scripts are run, spawning any process fails with `SystemResources` which usually means a file descriptor leak.

I'm able to reproduce this, but running `lsof -p <pid>` to list the currently open file descriptors shows only a handful at any given time. Not sure what's causing this issue. 

Interestingly, when running a large number of js files, one fd seems to be leaked per file that's executed. Not sure where those are coming 
(EDIT: there was a preexisting fd leak in run_command.zig that is now fixed. Error when just running `echo hi` persists).
(EDIT 2: Error when running `echo hi` is fixed)

> 3. it scans node_modules. It should not scan node_modules, .git, or any `.` directories.

Implemented by passing an ignore function to the `GlobWalker`
> 4. We need tests for what happens when you run multiple JavaScript/TypeScript files in different projects same process (`bun run --filter="*" index.js`). Currently, Bun assumes only one JavaScript file every executes for a given thread. That assumption is not true when using these code paths.

We now `fork` before executing any JS. This of course doesn't work on windows, and more generally we should decide how to handle this. Options include:
- Figure out the correct command line arguments and spawn a new copy of the bun process. Slowest, but would provide complete isolation. 
- Continue using fork and use something like `RtlCloneUserProcess` on windows. Caveat is that these are not "official" windows APIs and their stability is unknown.
- Work on isolating VM instances and run everything in the same process. Fastest, but also potentially risky?
> 5. It would be best to use the module resolution cache so that when running a JavaScript file later, the subsequent calls to getdents64() get cached. This will involve a small change to the GlobWalker.

Implemented by abstracting how the `GlobWalker` does filesystem access.

**Comment by gvilums** [comment]
After investigating 2. more, it seems like the error lies somewhere in child_process.spawnAndWait. Specifically, the error that's happening is that the child fails between being forked and calling execv, and signals this error to the parent. 

EDIT:
After switching the implementation to posix_spawn, the specific error that comes up seems to be E2BIG, which has a description of "The number of bytes in the new process's argument list is larger than the system-imposed limit.  This limit is specified by the sysctl(3) MIB variable KERN_ARGMAX." On my machine, this error starts occurring after spawning roughly 3000 processes. I'm not aware of any resource leaks that could cause this (after all, each process is spawned with an argument list that's basically identical, and each process has its own argument list), but I propose that we leave this issue be for now, as it is an extreme edge case

EDIT 2:
So I was able to solve the mystery: The culprit was an ever-growing path variable, which, each time an attempt at executing a package was made, was extended with new items. This lead to it growing unmanageably large after a few thousand executions. The fix was to reset PATH after every execution.

File: src/cli/run_command.zig
Comment (gvilums) [reply src/cli/run_command.zig]: The `cwd` member does seem to be used in the `kind` function [here](https://github.com/oven-sh/bun/blob/26a720defe3a0962f8b1062734b221b838613d4a/src/fs.zig#L1252-L1270), but that's the only spot. The other places in the `fs` implementation just use `std.fs.cwd`. So I think to ensure that the `kind` of a file is correctly resolved this should be updated (but maybe I'm wrong on that, in general the `cwd` handling is kind of a mess)

**Comment by MickL** [comment]
Does this also work with `bun add`? It is very annoying that I always need to go in the directory. pnpn has `pnp add --filter workspace`

File: src/glob.zig
Comment (dylan-conway) [inline src/glob.zig]: Is this `cwd` supposed to be used with `initWithCwd` on line 788?

File: src/deps/tinycc
Comment (dylan-conway) [inline src/deps/tinycc:1]: It looks like there isn't a diff, was this change intentional?

File: src/cli/filter_arg.zig
Comment (Jarred-Sumner) [inline src/cli/filter_arg.zig:69]: We are parsing the package.json files multiple times here since it may happen again later 

I'm hesitant to block the PR on this but we should figure out a way to cache this to avoid the overhead 

Also, we can simplify this code a bit:

`bun.sys.File.toSource(json_path, allocator);`

File: src/cli/filter_arg.zig
Comment (Jarred-Sumner) [inline src/cli/filter_arg.zig]: feels like this will be very slow but i guess the filters won't be long so it should be fine

File: src/cli/run_command.zig
Comment (Jarred-Sumner) [inline src/cli/run_command.zig]:
```suggestion
            // TODO: remember to free this when we add --concurrent
```

File: src/cli/filter_arg.zig
Comment (dylan-conway) [inline src/cli/filter_arg.zig]: we could skip `CMakeFiles` as well, but it's less likely to exist so probably not necessary

File: src/cli/filter_arg.zig
Comment (dylan-conway) [inline src/cli/filter_arg.zig]: We need an error message here

File: src/cli/run_command.zig
Comment (paperclover) [inline src/cli/run_command.zig]: on a different branch i had a version of this function that did one large allocation so that the defer free would be a single free.

code is https://github.com/oven-sh/bun/pull/9642 and might be cleaner to use.

it was put on pause because of change of focus to get 1.1 out.

File: src/cli/run_command.zig
Comment (paperclover) [reply src/cli/run_command.zig]: envp is freed. this entire comment can be removed now.

File: src/cli/bunx_command.zig
Comment (paperclover) [inline src/cli/bunx_command.zig]: this is expected to exit with the code the subprocess does,
and if the subprocess exits with a signal code instead of an exit code (SIGSEGV, SIGILL), we should raise the same one. the code here will only exit with 1.

i hope there are already tests for this but you can try to `bunx node -e 'process.abort()'`

**Review by Jarred-Sumner — COMMENTED** [review]
this needs more tests:
- what happens when a permissions error occurs when reading package.json or globbing? 
- what happens when a package.json is malformed?
- we should have a test that bun returns the appropriate exit code when a script run via --filter exits with a non-zero exit code
- we should look at tests from pnpm or yarn or npm to copy from
- a test that checks script & package names with non-ascii characters works

File: src/cli/run_command.zig
Comment (paperclover) [inline src/cli/run_command.zig]: can you add `signal: u32` here, this will make this similar to `std.process.Child.Term`. stopped and unknown are not as important.

and then i recommend adding some function in here to exit with the execresult, which will call exitWide for the codes, but use `raiseIgnoringPanicHandler` for signals. (posix only)

**Review by paperclover — CHANGES_REQUESTED** [review]

**Comment by dylan-conway** [comment]
> this needs more tests:
> 
> * what happens when a package.json is malformed?

Missing name and empty name are two tests we should have. Pnpm seems to not allow filtering with `--filter ""` so we can exit early when that pattern is given

**Comment by Jarred-Sumner** [comment]
> Missing name and empty name are two tests we should have. 

It's reading package.jsons to check the name/version in this code, so we need to test that here too

File: src/deps/tinycc
Comment (gvilums) [reply src/deps/tinycc:1]: nope, not intended

File: src/glob.zig
Comment (gvilums) [reply src/glob.zig]: I think this is leftover code, I removed the cwd stuff

File: src/cli/run_command.zig
Comment (gvilums) [reply src/cli/run_command.zig]: Done

File: src/cli/bunx_command.zig
Comment (gvilums) [reply src/cli/bunx_command.zig]: Makes sense, I updated the ExecResult to include an `exit` method which does the right thing

File: src/cli/filter_arg.zig
Comment (gvilums) [reply src/cli/filter_arg.zig]: Added

File: src/cli/filter_arg.zig
Comment (Jarred-Sumner) [reply src/cli/filter_arg.zig]: we should also skip`.git`

Comment (Jarred-Sumner) [reply src/cli/filter_arg.zig]: but that can be added later

**Comment by beorn** [comment]
Perhaps also support the '-F' shorthand?

**Comment by kevinmarrec** [comment]
@gvilums @Jarred-Sumner @dylan-conway 

Dependency order is not respected as said in the docs since this PR
https://github.com/oven-sh/bun/issues/12203
https://github.com/oven-sh/bun/issues/13239

I may have never worked, or there has been a breaking chance since.

**Pull request merged**
